lib/bcchelper.py

```diff
@@ -0,0 +1,636 @@
+and a required last member of cpuid.  Other key members can be added for the
+key types needed for that collector.  For example the structure for the
+default nfs collector has an op value to keep separate reads and write
+statistics.
+
+typedef struct {
+    u64  t;
+    char op[OP_NAME_LEN];
+    u32  cpuid;
+} nfs_key_t;
+
+The C HIST_KEY macro defines histogram key that is derived from the
+base aggregation key and adds a slot elemtent.  For example, the
+nfs_hist_key_t is define with the following macro:
+
+HIST_KEY(nfs_hist_key_t, nfs_key_t);
+
+This is the structure that is defined:
+
+typedef struct {
+    nfs_key_t agg_key;
+    u64       slot;
+} nfs_hist_key_t;
+
+See appliance/server/core/src/main/resources/analytics/nfs_layer_linux.st
+for an example of how to structure C bcc code.  In the python code a
+BPF object and an equality function for the key structure are required
+to initialize a bcc Helper.  The equality function should return true
+for keys that match in all member fields except the cpuid.  Aggregation
+and key type objects are added to the helper that describe what is being
+traced.  With this initialization, the helper printall method can be called
+repeatedly.  Each call will clear all the aggregation items since
+the last one and output the data.  The values for each timestamp are
+kept separate.  Within each timestamp the values from different
+cpus(cpuid) are combined for matching keys for each aggregation.
+IF ANALYTICS_PRINT_MODE is specified json objects are generated with
+the value of each aggregation for each timestamp suitable to be used
+by the appliance analytics system such as:
+
+{"t":"1565031672", "op":"write", "count":"92", "throughput":"645120", \
+"avgLatency":"609497", "latency":" \
+{500000,17},{600000,38},{700000,21},{800000,6},{900000,9},{2000000,1},"}
+
+The default output mode produces human readable output showing for the
+purpose of debugging:
+
+count[1565031672, write] = 92
+
+throughput[1565031672, write] = 645120
+
+avgLatency[1565031672, write] = 609497
+
+latency[1565031672, write] =
+500000		17
+600000		38
+700000		21
+800000		6
+900000		9
+2000000		1
+"""
+
+
+class BCCHelper:
+    SUM_AGGREGATION = 0
+    COUNT_AGGREGATION = 1
+    AVERAGE_AGGREGATION = 2
+    AVERSUM_AGGREGATION = 3
+    STDDEV_AGGREGATION = 4
+    # Logic in isHistogram() requires Histogram values > other aggregations
+    LOG_HISTOGRAM_AGGREGATION = 5
+    LL_HISTOGRAM_AGGREGATION = 6
+    DEFAULT_PRINT_MODE = 0
+    ANALYTICS_PRINT_MODE = 1
+    ESTAT_PRINT_MODE = 2
+    #
+    # For each key type, python string post-processing can be added in
+    # key_type_string(); much easier than in the bcc C code.
+    #
+    DEFAULT_KEY_TYPE = 0
+    IP_KEY_TYPE = 1
+    # Format string for print histograms
+    ESTAT_HIST_FORMAT = '{:<16} {:>16} {:<41}'
+    ESTAT_HEADER_FORMAT = '   {:<20}{:>47}'
+    ESTAT_SCALAR_FORMAT = '{:>12}'
+    ESTAT_TITLE_FORMAT = '{:<30}'
```

Comment:
If we change the name we will want to make sure we change these too and anywhere that `estat` or `ESTAT` is used.

Reply:
Do we have agreement on a name?

Reply:
It is still estat in Prashanth's PR.   Agreed that we should change it everywhere if we change it.